interp/iostream_test.go

```diff
@@ -0,0 +1,321 @@
+	shell := fmt.Sprintf("exit %d", exitCode)
+	return shellTest{
+		Name:     "ExitCode",
+		shell:    shell,
+		exitCode: exitCode,
+		output:   emptyOutput(),
+	}
+}
+
+func randSignalTest() shellTest {
+	signal := 1 + rand.Intn(15)
+	shell := fmt.Sprintf("exec kill -%d $$", signal)
+	return shellTest{
+		Name:     "ExitSignal",
+		shell:    shell,
+		exitCode: 256 + signal,
+		output:   emptyOutput(),
+		requires: "kill",
+		skipWin:  true,
+	}
+}
+
+func TestStreamInCmd(t *testing.T) {
+	t.Run("ShellMissing", func(t *testing.T) {
+		fname := randMissingExe()
+		in, err := inputCmdStream(exec.Command(fname))
+		if in != nil {
+			t.Error("should get nil stream on missing command")
+		}
+		checkShellMissing(t, err)
+	})
+
+	for _, tc := range shellTestCases {
+		t.Run(tc.Name, tc.TestInputShell)
+	}
+}
+
+func TestStreamOutCmd(t *testing.T) {
+	t.Run("ShellMissing", func(t *testing.T) {
+		fname := randMissingExe()
+		out, err := outputCmdStream(exec.Command(fname))
+		if out != nil {
+			t.Error("should get nil stream on missing command")
+		}
+		checkShellMissing(t, err)
+	})
+	for _, tc := range shellTestCases {
+		t.Run(tc.Name, tc.TestOutputShell)
+	}
+}
+
+func checkShellMissing(t *testing.T, err error) {
+	if _, ok := err.(*exec.Error); !ok {
+		t.Errorf("error should be *exec.Error, got: %#v", err)
+	}
+	if !strings.Contains(err.Error(), "executable file not found in $PATH") {
+		t.Error("expected error containing \"executable file not found in $PATH\"")
+	}
+}
+
+func checkExitCode(t *testing.T, expectCode, exitCode int) {
+	if exitCode != expectCode {
+		t.Errorf("expected Close() result: %d, got: %d", expectCode, exitCode)
+	}
+}
+
+// Generate a random executable name this is guaranteed not to exist on the $PATH.
+func randMissingExe() string {
+	for i := 0; ; i++ {
+		if i >= maxRandExeAttempts {
+			panic("failed to generate random missing executable")
+		}
+		buf := make([]byte, 2)
+		if _, err := rand.Read(buf); err != nil {
+			panic(err)
+		}
+		fname := fmt.Sprintf("noshell%04X", buf)
+		if path, _ := exec.LookPath(fname); path == "" {
+			return fname
+		}
+	}
+}
+
+// Skip the test if any of the executables in exes are not available on the $PATH. exes is a space
+// delimited string of executable names.
+func requireExe(t *testing.T, exes string) {
+	for _, exe := range strings.Split(exes, " ") {
+		if exe == "" {
+			continue
+		}
+		if path, _ := exec.LookPath(exe); path == "" {
+			t.Skipf("test case requires %s", exe)
+		}
+	}
+}
+
+// Output checkers are functions which spawn goroutines to read lines from the io.ReaderCloser and
+// check the input. After the content passes or fails the check, they Close() the io.ReadCloser.
+// They send two errors: the first is any problem with the content of the input (or nil) and the second
+// is the error result of Close(). These functions return output checkers as closures.
+
+type outputChecker func(rc io.ReadCloser) chan error
+
+func emptyOutput() outputChecker {
```

Comment:
I *think* all of this concurrency/complexity will go away if we can use regular interp tests, as the interpreter handles the command concurrency already.

Reply:
Ok, removed.